File: python/src/d04.py
"""
Advent Of Code 2022 Day 4
"""

from __future__ import annotations


from pathlib import Path

import utils


def range_to_set(range_str: str) -> set[int]:
    start, end = range_str.split("-")
    return set(range(int(start), int(end) + 1))


def p1p2(input_file: Path = utils.real_input()) -> tuple[int, int]:
    p2 = 0
    num_subset_pairs = 0
    for line in input_file.read_text().splitlines():
        elf1, elf2 = line.split(",")

        elf1_set = range_to_set(elf1)
        elf2_set = range_to_set(elf2)
        if elf1_set.issubset(elf2_set) or elf2_set.issubset(elf1_set):
            num_subset_pairs += 1
        if not elf1_set.isdisjoint(elf2_set):
            p2 += 1

    return (num_subset_pairs, p2)

File: rust/src/d04.rs
use itertools::Itertools;

use crate::utils::*;

pub fn p1p2(input_file: &str) -> AoCSolver {
    let mut p1: i64 = 0;
    let mut p2: i64 = 0;

    for line in read_lines(input_file).unwrap() {
        let line = line.unwrap();
        let (elf1, elf2) = line.split(",").next_tuple().unwrap();
        let (elf1_start, elf1_end) = range_str_parse(elf1);
        let (elf2_start, elf2_end) = range_str_parse(elf2);

        if elf1_start <= elf2_start && elf1_end >= elf2_end ||
           elf2_start <= elf1_start && elf2_end >= elf1_end {
            p1 += 1;
        }
        if elf1_end >= elf2_start && elf1_start <= elf2_end ||
           elf2_end >= elf1_start && elf2_start <= elf1_end {
            p2 += 1;
        }

    }

    AoCSolver::BothParts(AoCResult::Number(p1),
                         AoCResult::Number(p2))
}

fn range_str_parse(range: &str) -> (i64, i64) {
    range.split("-")
            .map(|x| x.parse::<i64>().unwrap())
            .collect_tuple().unwrap()
}